### DISCUSSION TITLE
TMC Location tables question

### DISCUSSION THREAD

**Discussion by veso266** [windytan/redsea#98] [General]
Sorry to bother you, but I just don't know where else to ask this question
I am playing with a map for VW/seat/Skoda cars (RNS510), trying to add our Slovenian TMC location tables to the map

Just wondering if anyone maybe saw this file format before

TMC consists of TMCINFO.ini which looks like this
```
[PROD]
PROD =  "RNS_EU_38"
SCOPE = "EU"
; DATE "2019_06_07H16:56"
;IoCo: Updated Austria, Belgium, Finland, France, Germany, Italy, Norway, Switzerland, UK with new versions.
[LT]
AUSTRIA        = 0xa,  1, "AT",   "aut_A_1_3_4",    9.53485,   46.41923,   17.11080,   48.99898, 0,       272678, , "A"
BELGIUM        = 0x6,  1, "BE",   "bel_6_1_3_1",    2.55995,   49.50883,    6.34339,   51.48590, 0,       187900, , "B"
SWITZERLAND    = 0x4,  9, "CH",   "che_4_9_6_7",    5.96445,   45.64020,   10.58965,   48.00405, 0,       167488, , "CH"
CZECH_REPUBLIC = 0x2, 25, "CZ",  "cze_2_25_7_0",   12.16019,   48.56245,   18.85167,   51.02279, 0,       407735, , "CZ"
GERMANY1       = 0x1,  1, "DE",  "deu_D_1_18_0",    5.87690,   47.38710,   15.00870,   55.01726  0,       857037, , "D"
GERMANY2       = 0xd,  1, "DE",  "deu_D_1_18_0",    5.87690,   47.38710,   15.00870,   55.01726, 0,       857037, , "D" 
DENMARK        = 0x9,  9, "DK",  "dnk_9_9_12_0",    7.91384,   54.18433,   15.14151,   59.18825, 0,       229504, , "DK"
SPAIN          = 0xe, 17, "ES",  "esp_E_17_4_1",  -16.74393,   27.75962,    3.18844,   43.73141, 0,       593935, , "E"
FINLAND        = 0x6, 17, "FI",  "fin_6_17_2_5",   19.00100,   59.82831,   31.48623,   70.35791, 0,       668423, , "FIN"  
FRANCE         = 0xf, 32, "FR", "fra_F_32_12_1",   -4.77149,   42.36702,    8.18458,   51.06969, 0,       950911, , "F"
UNITED_KINGDOM1= 0xc,  7, "GB",   "gbr_C_7_4_1",   -8.09545,   43.33306,    8.43887,   60.48963, 0,       519343, , "GB"
ITALY          = 0x5,  1, "IT",   "ita_5_1_4_3",    6.67319,   36.69411,   18.51336,   47.04999, 0,       922643, , "I"
NETHERLANDS    = 0x8, 17, "NL",  "nld_8_17_9_5",    3.35977,   50.75458,    7.22918,   53.44311, 0,       200305, , "NL"
NORWAY         = 0xf, 49, "NO",  "nor_F_49_6_0",    4.67033,   57.98448,   31.11448,   71.16779, 0,       320330, , "N"
SWEDEN         = 0xe, 33, "SE",  "swe_E_33_3_2",   11.15119,   55.34201,   24.14387,   68.44281, 0,       366049, , "S"
[LTX]
CZEECH_cs_utf8  = 0x2, 25, "CZ", "cze_2_25_7_0_1",   12.16019,   48.56245,   18.85167,   51.02279, 0,       425838, , "CZ", "cs"
[ET]
DANISH      = "dk", "danish",     "dansk",      0, 14413,
DUTCH       = "nl", "dutch",      "nederlands", 0, 18902,
ENGLISH     = "en", "english",    "english",    0, 13200,
FINNISH     = "fi", "finnish",    "suomi",      0, 14332,
FLEMISH     = "nl-BE", "flemish",    "vlaams",  0, 20388,
FRENCH      = "fr", "french",     "francais",   0, 14454,
GERMAN      = "de", "german",     "deutsch",    0, 10605,
ITALIAN     = "it", "italian",    "italiano",   0, 15510,
PORTUGUESE  = "pt", "portuguese", "portuguese", 0, 15231,
SPANISH     = "es", "spanish",    "espanol",    0, 12908,
CZECH       = "cs", "czech",      "czech",      0, 23183,
HUNGARIAN   = "hu", "hungarian",  "hungarian",  0, 16437,
NORWEGIAN   = "no", "norwegian",  "norwegian",  0, 10866,
POLISH      = "pl", "polish",     "polish",     0, 12236,
RUSSIAN     = "ru", "russian",    "russian",    0, 17280,
SWEDISH     = "sv", "swedish",    "svenska",    0, 13619,
TURKISH     = "tr", "turkish",    "turkish",    0, 11968,
CHINESE     = "zh", "chinese",    "chinese",    0, 16716,
[CAT]
CAT = "eventinfo_2011_11_30_eu_nar_chnexc"
[SETTINGS]
[ACP]
F-VM = "@Ps|YudMWEEGSxx|"
```
Looking at this file, it appears that 
[LT] references files inside LOC folder and 
[ET] references files inside LAN folder

all the files have readable strings inside

This file looks very familiar to me for some reason, I just don't know what it belongs to

I also have the program that supposebly reads this file format called DBAL.OUT (sadly only in binary form, and to make situation worse, the program is supposed to run on VxWorks (although if I am not mistaken, thats similar to linux, since it supports POSIX commands)

So yea, I am just wondering if anyone, maybe, just maybe would know this format

[DBAL.zip](https://github.com/windytan/redsea/files/12643569/DBAL.zip)
[TMC2.zip](https://github.com/windytan/redsea/files/12643551/TMC2.zip)

I know this is totaly not the right place to ask this thing, but not realy sure where would I, since you know TMC better then anyone else

**Comment by windytan** [comment]
I'll actually move this to the newly created [Discussions](https://github.com/windytan/redsea/discussions/) section, then it's exactly in the right place!

Personally I haven't seen this kind of a binary location table format before. Normally LT's are in the "TMC Exchange" (CSV-based) format. This could also be their own proprietary database format.

**Comment by andimik** [comment]
Sorry dear @veso266 for the late answer. 

I'm living in Austria near the Slovenian border and checked the Slovenian Location table, which I found on https://ncup.si/en/loctable 

But I get clear data (checked my complete archive) out, so I wonder why you have problems. Note, you need to tune to the first or second public channel (Prvi and Val 202)

```
cat /media/andreas/Data/ut/rdsspy/9*.spy | redsea -h -l /media/andreas/Data/ut/rdsspy/locationcode/aut/AT_E0_A_1_3.8_LTEF -l /media/andreas/Data/ut/rdsspy/locationcode/slo/ | jq ".tmc.message.road_name" | grep -v null | sort | uniq | head
"Ajdovščina - Nova Gorica"
"Ajševica - Nova Gorica"
"Bertoki - Gračišče"
"Bevško - most čez Savo"
"Bistrica pri Tržiču - Kranj"
"Bitnje - Stara Fužina"
"Bizeljsko - Krško"
"Bled - Bohinj"
"Bohinj - Tolmin"
"Borjana - Robidišče"
```

Hint: for **Austria**, the location table can be obtained free of charge from https://contentportal.asfinag.at/location-code-list/terms if you are interested in as well.

```
cat /media/andreas/Data/ut/rdsspy/A*.spy | redsea -h -l /media/andreas/Data/ut/rdsspy/locationcode/aut/AT_E0_A_1_3.8_LTEF -l /media/andreas/Data/ut/rdsspy/locationcode/slo/  | jq ".tmc.message.road_name" | grep -v null | sort | uniq
"Aggsteiner Straße"
"Alser Straße"
"Altmannsdorfer Straße"
"Arlberg Schnellstraße"
"Arlbergstraße"
"Attersee Straße"
```

For **Germany**: https://www.bast.de/DE/Verkehrstechnik/Fachthemen/v2-LCL/location-code-list.html

```
$ cat /media/andreas/Data/ut/rdsspy/rds-spy-logs/D*.spy | redsea -h -l /media/andreas/Data/ut/rdsspy/locationcode/aut/AT_E0_A_1_3.8_LTEF -l /media/andreas/Data/ut/rdsspy/locationcode/slo/ -l /media/andreas/Data/ut/rdsspy/locationcode/de/LCL22.0.D_230515/ | jq ".tmc.message.road_name" | grep -v null | sort | uniq
"Colditzer Straße"
"Dresdner Straße"
"Ernst Reuter Allee"
"Kölner Ring"
"L 1112"
"Leipziger Chaussee"
"Merseburger Landstraße"
"Mittlerer Ring"
"Pariser Straße"
"Rheinallee"
"Südtangente"
```

**Reply by veso266** [reply]
Not a problem on ur late response, some things just take time

For instance (not related), in 20years I still havent able to find local box (a service device) that would allow me to program new phone number and see diagnostic data about my Benefon Forte TDM-10J: 

https://youtu.be/nX9Jo9XX5F8

I slowly collect manuals and souch here: http://radio1.dxing.si/Telephone/NMT%20%28Nordic%20Mobile%20Telephone%29/

But a diagnostic service box, nowhere to be found (cant even find picture or anyinfo about it online)

Well the thing is that maps for RNS510 navigation system (included in 2010 in VW, Škoda, Seat) are not made anymore and Slovenian location tables were never included (not even TrafficNav (payed) ones that are broadcasted on Radio  SI), so no traffic info for Slovenia

And I had an idea to try to inject the free ones into my maps so I could have news about accidents and roadworks and souch

Of course later I wanted to make new maps using OpenStreetMaps (but injecting my own tmc location tables seamed to be easier step)

**Comment by andimik** [comment]
hint to @windytan : in https://github.com/windytan/redsea/wiki/Supported-RDS-features  `road_name` is not mentioned.

**Comment by mvglasow** [comment]
Pretty sure it’s some proprietary, vendor-specific binary format, lest someone get the urge to tinker.

Some location tables are proprietary and available only in exchange for royalty payments. Manufacturers who use these location tables have received them under a licensing agreement, which might require them to take measures against these tables being reused (as in, another product for which the licensor is not getting any royalty payments).

Some manufacturers might also wish to make you pay for upgrades rather than hack up your own location tables.

And finally, even location tables which are available free of charge still frequently come with restrictions, including on redistribution. At least one has licensing terms which state that the LT may only be distributed embedded in a product, not on its own. Makes me wonder whether just converting the individual tables into a SQLite database and shipping an app with the DB satisfies that requirement.

If you are interested in plain-text location tables available free of charge, check out https://wiki.openstreetmap.org/wiki/TMC#Available_datasets. Some can just be downloaded, others require you to agree to licensing terms – but still much less hassle than extracting binary files from your sat-nav and figuring out how to decode them.

**Reply by veso266** [reply]
Thnx, for free ones, sure its easier to downloaf but for TrafficNav ones the only option is extraxtion

Also maybe u would want to inject free ones into navigation that didnt bother to include ur country ones (this happened a lot, Italian, Austrian ones, included, while Slovenian are missing, even if they are free of charge)

**Reply by andimik** [reply]
Thanks for the link, I have visited it in the past (when redsea started to support TMC), but could not find it this week via the search engine.

**Comment by mvglasow** [comment]
It might not be as simple as just injecting a bunch of files. For example, in the TMC location table, a junction is just a point at a given coordinate pair. However, when this junction appears in a TMC message, the actual coordinates are slightly off: depending on whether the obstruction begins or ends at that junction, the actual coordinates used are those where the exit ramp leaves the main carriageway, or where the entry ramp joins it. On multi-carriageway roads, coordinates also depend on the carriageway, i.e. the direction. Also, each map is slightly different. So TMC data has to be adapted to the map, which may require hand-massaging of data. (Sure, there are ways to automate this, but they don’t work in every case.)
